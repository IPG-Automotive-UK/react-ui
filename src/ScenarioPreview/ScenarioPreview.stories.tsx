import { Meta, StoryFn } from "@storybook/react";

import React from "react";
import { ScenarioPreview } from "./ScenarioPreview";
import { ScenarioPreviewProps } from "./ScenarioPreview.types";

const meta: Meta<typeof ScenarioPreview> = {
  component: ScenarioPreview,
  title: "Preview/ScenarioPreview"
};
export default meta;

// Story Template
const Template: StoryFn<ScenarioPreviewProps> = args => {
  return <ScenarioPreview {...args} />;
};

/**
 * Default story
 */
export const Default = Template.bind({});
Default.args = {
  createdAt: "23-06-06 08:09:24",
  description: "Small Description",
  file: "RoadName",
  format: "CarMaker",
  formatVersion: "11.1",
  href: "https://example.com",
  image: "https://picsum.photos/id/191/400/200",
  name: "CityDriving",
  roadHref: "https://example.com/road-long",
  roadName: "A very long road name that might overflow",
  sx: {
    borderRadius: 1,
    boxShadow: 3,
    maxWidth: "480px",
    padding: 2
  },
  user: { name: "James Harper" }
};

/**
 * Story with long text to test overflow handling
 */
export const WithOverflowText = Template.bind({});
WithOverflowText.args = {
  ...Default.args,
  description:
    "This is a very long description that might cause overflow issues if not handled properly. The description highlights a scenario with intricate details about the simulation.",
  file: "A_very_very_very_very_very_very_long_long_long_long_long_long_long_long_long_long_long_long_long_long_long_long_scenario_file_name_that_exceeds_typical_length.scn",
  name: "A very very very very very very very very very very very long long long long long scenario name that is unlikely to fit in a single line.",
  roadHref:
    "https://example.com/road-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long-long",
  roadName:
    "A very long road long road long road long road long road name that might overflow",
  user: {
    color: "rgb(236, 64, 122)",
    name: "A very long user name that might overflow the container"
  }
};

/**
 * Story without optional properties
 */
export const WithoutOptionalProps = Template.bind({});
WithoutOptionalProps.args = {
  description: "A minimal scenario description.",
  file: "MinimalFile.scn",
  format: "Format",
  formatVersion: "1.0",
  href: "https://example.com",
  image: "https://picsum.photos/id/514/400/200",
  name: "Minimal Scenario",
  roadHref: "https://example.com/minimal-road",
  roadName: "Minimal Road",
  sx: {
    borderRadius: 1,
    boxShadow: 3,
    maxWidth: "480px",
    padding: 2
  }
};

/**
 * Story with custom labels
 */
export const WithLabels = Template.bind({});
WithLabels.args = {
  ...Default.args,
  label: [
    { color: "#FFC107", name: "Label 1" },
    { color: "#4CAF50", name: "Label 2" }
  ]
};
