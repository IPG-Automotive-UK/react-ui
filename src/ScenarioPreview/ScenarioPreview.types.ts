import { SxProps, Theme } from "@mui/material";

import { Label } from "../LabelSelector/Label.types";

/**
 * Type of user associated with the scenario
 */
export type User = {
  /**
   * Name of the user
   */
  name: string;
  /**
   * Optional color for the user's avatar
   */
  color?: string;
};

/**
 * The props needed to define a scenario preview
 */
export type ScenarioPreviewProps = {
  /**
   * Name of the scenario
   */
  name: string;
  /**
   * Link to open the current scenario
   */
  href: string;
  /**
   * Image URL of the scenario
   */
  image: string;
  /**
   * Description of the scenario
   */
  description: string;
  /**
   * Format of the scenario (e.g., ASAM OpenDRIVE)
   */
  format: string;
  /**
   * Format version of the scenario
   */
  formatVersion: string;
  /**
   * File object representing the scenario file
   */
  file: string;
  /**
   * Date of creation of the scenario in ISO-8601 format
   */
  createdAt?: string;
  /**
   * Optional user who created the scenario
   */
  user?: User;
  /**
   * Optional labels associated with the scenario
   */
  label?: Label[];
  /**
   * Name for the road
   */
  roadName: string;
  /**
   * Link to open the current road
   */
  roadHref: string;
  /**
   * Additional styles to apply to the scenario preview wrapper
   */
  sx?: SxProps<Theme>;
};
