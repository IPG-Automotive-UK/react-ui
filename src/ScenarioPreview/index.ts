export { ScenarioPreview } from "./ScenarioPreview";
export type { ScenarioPreviewProps } from "./ScenarioPreview.types";
