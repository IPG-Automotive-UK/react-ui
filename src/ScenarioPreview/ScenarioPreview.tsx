import { Box, Divider, Link, Stack, Typography } from "@mui/material";
import { LabelChipGroup, LabelChipGroupProps } from "../LabelSelector";

import DateLabel from "../DateLabel/DateLabel";
import FileLabel from "../FileLabel/FileLabel";
import FormatLabel from "../FormatLabel/FormatLabel";
import FormatVersionLabel from "../FormatVersionLabel/FormatVersionLabel";
import NoWrapTypography from "../NoWrapTypography/NoWrapTypography";
import React from "react";
import RoadLabel from "../RoadLabel/RoadLabel";
import { ScenarioPreviewProps } from "./ScenarioPreview.types";
import UserLabel from "../UserLabel/UserLabel";

export function ScenarioPreview({
  name,
  href,
  image,
  description,
  format,
  formatVersion,
  file,
  createdAt,
  user,
  label = [],
  roadLabel,
  sx
}: ScenarioPreviewProps) {
  function hasOptionalProperty() {
    return label?.length > 0 || createdAt || user;
  }

  const labelChips: LabelChipGroupProps["chips"] = label?.map(l => ({
    clickable: false,
    color: l.color,
    label: l.name,
    size: "small"
  }));

  return (
    <Box
      display="flex"
      flexDirection="column"
      gap={1}
      minWidth={0}
      fontFamily="Montserrat"
      data-testid="scenario-preview-wrapper"
      sx={{ ...sx }}
    >
      <Box>
        <Stack direction="row" spacing={1} minWidth={0}>
          <img
            src={image}
            alt="scenario-image"
            style={{
              height: "44px",
              marginBottom: "auto",
              marginTop: "auto",
              objectFit: "contain",
              width: "78px"
            }}
          />
          <Stack direction="column" minWidth={0}>
            <NoWrapTypography>
              <Link
                href={href}
                target="_blank"
                color="primary"
                variant="subtitle2"
                underline="hover"
                data-testid="scenario-preview-name"
              >
                {name}
              </Link>
            </NoWrapTypography>

            <Typography
              color="textPrimary"
              data-testid="scenario-preview-description"
              sx={{
                WebkitBoxOrient: "vertical",
                WebkitLineClamp: 2,
                display: "-webkit-box",
                overflow: "hidden",
                textOverflow: "ellipsis"
              }}
              variant="caption"
            >
              {description}
            </Typography>
          </Stack>
        </Stack>
      </Box>
      <Box>
        <Stack
          direction="row"
          gap={"12px"}
          justifyContent="left"
          alignItems="center"
        >
          <Box data-testid="format-label" flex="0 1 auto">
            <FormatLabel label={format} />
          </Box>
          <Box data-testid="format-version-label" flex="0 1 auto">
            <FormatVersionLabel label={formatVersion} />
          </Box>
          <Box
            data-testid="file-label"
            flex="0 1 auto"
            maxWidth={"calc(40% - 12px)"}
            alignItems="center"
          >
            <FileLabel label={file} />
          </Box>
          {roadLabel && (
            <Box data-testid="road-label">
              <RoadLabel label={roadLabel} href={href} />
            </Box>
          )}
        </Stack>
      </Box>
      {hasOptionalProperty() && (
        <>
          <Divider />
          <Box display="flex" flexDirection="column" gap={1}>
            <Stack direction="row" gap={"12px"}>
              {createdAt && (
                <Box data-testid="date-label" flex="0 1 auto">
                  <DateLabel label={createdAt} />
                </Box>
              )}
              {user && (
                <Box
                  data-testid="user-label"
                  flex="1 1 auto"
                  alignItems="center"
                  minWidth={0}
                >
                  <UserLabel label={user.name} color={user.color} />
                </Box>
              )}
            </Stack>
            {label?.length > 0 && (
              <Stack direction="row" spacing={1}>
                <LabelChipGroup chips={labelChips} />
              </Stack>
            )}
          </Box>
        </>
      )}
    </Box>
  );
}
