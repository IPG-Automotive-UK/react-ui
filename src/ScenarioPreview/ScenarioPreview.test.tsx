import {
  Default,
  WithOverflowText,
  WithoutOptionalProps
} from "./ScenarioPreview.stories";
import { render, screen } from "@testing-library/react";

import React from "react";
import { ScenarioPreview } from "./ScenarioPreview";

describe("ScenarioPreview", () => {
  // Test that the component renders
  test("renders without crashing", () => {
    render(
      <ScenarioPreview
        name={Default.args.name}
        href={Default.args.href}
        image={Default.args.image}
        description={Default.args.description}
        format={Default.args.format}
        formatVersion={Default.args.formatVersion}
        file={Default.args.file}
        createdAt={Default.args.createdAt}
        user={Default.args.user}
        roadLabel={Default.args.roadLabel}
      />
    );

    expect(screen.getByTestId("scenario-preview-wrapper")).toBeInTheDocument();
  });

  // Test that the road label renders correctly
  test("renders road label correctly", () => {
    render(
      <ScenarioPreview
        name={Default.args.name}
        href={Default.args.href}
        image={Default.args.image}
        description={Default.args.description}
        format={Default.args.format}
        formatVersion={Default.args.formatVersion}
        file={Default.args.file}
        roadLabel="Test Road Name"
      />
    );

    const roadLabelElement = screen.getByTestId("road-label");
    expect(roadLabelElement).toBeInTheDocument();
    expect(roadLabelElement).toHaveTextContent("Test Road Name");
  });

  // Test that the divider renders when optional props are present
  test("renders divider when optional props are provided", () => {
    const { container } = render(
      <ScenarioPreview
        name={Default.args.name}
        href={Default.args.href}
        image={Default.args.image}
        description={Default.args.description}
        format={Default.args.format}
        formatVersion={Default.args.formatVersion}
        file={Default.args.file}
        createdAt={Default.args.createdAt}
        user={Default.args.user}
        roadLabel={Default.args.roadLabel}
      />
    );

    expect(container.querySelector(".MuiDivider-root")).toBeInTheDocument();
  });

  // Test that no divider renders when optional props are not provided
  test("does not render divider when optional props are missing", () => {
    const { container } = render(
      <ScenarioPreview
        name={WithoutOptionalProps.args.name}
        href={WithoutOptionalProps.args.href}
        image={WithoutOptionalProps.args.image}
        description={WithoutOptionalProps.args.description}
        format={WithoutOptionalProps.args.format}
        formatVersion={WithoutOptionalProps.args.formatVersion}
        file={WithoutOptionalProps.args.file}
      />
    );

    expect(container.querySelector(".MuiDivider-root")).not.toBeInTheDocument();
  });

  // Test for handling overflow text in name, description, and road label
  test("handles overflow text correctly", () => {
    render(
      <ScenarioPreview
        name={WithOverflowText.args.name}
        href={WithOverflowText.args.href}
        image={WithOverflowText.args.image}
        description={WithOverflowText.args.description}
        format={WithOverflowText.args.format}
        formatVersion={WithOverflowText.args.formatVersion}
        file={WithOverflowText.args.file}
        roadLabel={WithOverflowText.args.roadLabel}
      />
    );

    expect(screen.getByText(WithOverflowText.args.name)).toBeInTheDocument();
    expect(
      screen.getByText(WithOverflowText.args.description)
    ).toBeInTheDocument();
    expect(screen.getByText(WithOverflowText.args.file)).toBeInTheDocument();
    expect(
      screen.getByText(WithOverflowText.args.roadLabel)
    ).toBeInTheDocument();
  });

  // Test that all properties are rendered correctly
  test("renders all provided properties correctly", () => {
    render(
      <ScenarioPreview
        name={Default.args.name}
        href={Default.args.href}
        image={Default.args.image}
        description={Default.args.description}
        format={Default.args.format}
        formatVersion={Default.args.formatVersion}
        file={Default.args.file}
        createdAt={Default.args.createdAt}
        user={Default.args.user}
        roadLabel={Default.args.roadLabel}
      />
    );

    const nameElement = screen.getByRole("link", { name: Default.args.name });
    const descriptionElement = screen.getByTestId(
      "scenario-preview-description"
    );
    const formatElement = screen.getByTestId("format-label");
    const formatVersionElement = screen.getByTestId("format-version-label");
    const fileElement = screen.getByTestId("file-label");
    const roadLabelElement = screen.getByTestId("road-label");
    const createdAtElement = screen.getByTestId("date-label");
    const userElement = screen.getByTestId("user-label");

    expect(nameElement).toHaveTextContent(Default.args.name);
    expect(descriptionElement).toHaveTextContent(Default.args.description);
    expect(formatElement).toHaveTextContent(Default.args.format);
    expect(formatVersionElement).toHaveTextContent(Default.args.formatVersion);
    expect(fileElement).toHaveTextContent(Default.args.file);
    expect(roadLabelElement).toHaveTextContent(Default.args.roadLabel);
    expect(createdAtElement).toHaveTextContent(Default.args.createdAt);
    expect(userElement).toHaveTextContent(Default.args.user.name);
  });
});
