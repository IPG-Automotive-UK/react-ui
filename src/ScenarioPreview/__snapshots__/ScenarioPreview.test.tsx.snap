// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScenarioPreview > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1s0ioqa"
    data-testid="scenario-preview-wrapper"
  >
    <div
      class="MuiBox-root css-1r99qxv"
    >
      <div
        class="MuiStack-root css-1vkv85v-MuiStack-root"
      >
        <img
          alt="scenario-image"
          src="https://picsum.photos/id/101/400/200"
          style="height: 44px; margin-bottom: auto; margin-top: auto; object-fit: contain; width: 78px;"
        />
        <div
          class="MuiStack-root css-6lanpv-MuiStack-root"
        >
          <div
            class="MuiStack-root css-zpgd58-MuiStack-root"
          >
            <div
              class="MuiBox-root css-0"
            />
            <p
              class="MuiTypography-root MuiTypography-body1 MuiBox-root css-1gcgdf8-MuiTypography-root"
              data-mui-internal-clone-element="true"
            >
              <a
                class="MuiTypography-root MuiTypography-subtitle2 MuiLink-root MuiLink-underlineHover css-1atbtaw-MuiTypography-root-MuiLink-root"
                data-testid="scenario-preview-name"
                href="https://example.com"
                target="_blank"
              >
                718-2024-CaymanS
              </a>
            </p>
          </div>
          <div
            class="MuiStack-root css-m69qwo-MuiStack-root"
          >
            <span
              class="MuiBox-root css-12d10d6"
              data-mui-internal-clone-element="true"
            >
              <span
                class="MuiTypography-root MuiTypography-caption css-x3aacn-MuiTypography-root"
                data-testid="scenario-preview-description"
              >
                Small Description
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1r99qxv"
    >
      <div
        class="MuiStack-root css-1grjenu-MuiStack-root"
      >
        <div
          class="MuiBox-root css-1afnuq6"
          data-testid="format-label"
        >
          <div
            class="MuiStack-root css-98m5zp-MuiStack-root"
            data-testid="icon-with-label"
          >
            <div
              class="MuiBox-root css-k008qs"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18r7n60-MuiSvgIcon-root"
                data-testid="default-format-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 4H4c-1.11 0-2 .9-2 2v12c0 1.1.89 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.89-2-2-2m0 14H4V8h16zm-2-1h-6v-2h6zM7.5 17l-1.41-1.41L8.67 13l-2.59-2.59L7.5 9l4 4z"
                />
              </svg>
            </div>
            <div
              class="MuiBox-root css-31l7gp"
            >
              <p
                class="MuiTypography-root MuiTypography-caption MuiTypography-noWrap css-1m86euf-MuiTypography-root"
              >
                AnyRandom Format
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1c8o7mf"
          data-testid="format-version-label"
        >
          <div
            class="MuiStack-root css-98m5zp-MuiStack-root"
            data-testid="icon-with-label"
          >
            <div
              class="MuiBox-root css-k008qs"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18r7n60-MuiSvgIcon-root"
                data-testid="format-version-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m20.5 10 .5-2h-4l1-4h-2l-1 4h-4l1-4h-2L9 8H5l-.5 2h4l-1 4h-4L3 16h4l-1 4h2l1-4h4l-1 4h2l1-4h4l.5-2h-4l1-4zm-7 4h-4l1-4h4z"
                />
              </svg>
            </div>
            <div
              class="MuiBox-root css-31l7gp"
            >
              <p
                class="MuiTypography-root MuiTypography-caption MuiTypography-noWrap css-1m86euf-MuiTypography-root"
              >
                1.0
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1afnuq6"
          data-testid="file-label"
        >
          <div
            class="MuiStack-root css-98m5zp-MuiStack-root"
            data-testid="icon-with-label"
          >
            <div
              class="MuiBox-root css-k008qs"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18r7n60-MuiSvgIcon-root"
                data-testid="file-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.5 6v11.5c0 2.21-1.79 4-4 4s-4-1.79-4-4V5c0-1.38 1.12-2.5 2.5-2.5s2.5 1.12 2.5 2.5v10.5c0 .55-.45 1-1 1s-1-.45-1-1V6H10v9.5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V5c0-2.21-1.79-4-4-4S7 2.79 7 5v12.5c0 3.04 2.46 5.5 5.5 5.5s5.5-2.46 5.5-5.5V6z"
                />
              </svg>
            </div>
            <div
              class="MuiBox-root css-31l7gp"
            >
              <p
                class="MuiTypography-root MuiTypography-caption MuiTypography-noWrap css-1m86euf-MuiTypography-root"
              >
                TestScenarioFile.scn
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-yd8sa2"
    >
      <div
        class="MuiStack-root css-jy6k46-MuiStack-root"
      >
        <div
          class="MuiBox-root css-1afnuq6"
          data-testid="date-label"
        >
          <div
            class="MuiStack-root css-98m5zp-MuiStack-root"
            data-testid="icon-with-label"
          >
            <div
              class="MuiBox-root css-k008qs"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18r7n60-MuiSvgIcon-root"
                data-testid="date-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M9 11H7v2h2zm4 0h-2v2h2zm4 0h-2v2h2zm2-7h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2m0 16H5V9h14z"
                />
              </svg>
            </div>
            <div
              class="MuiBox-root css-31l7gp"
            >
              <p
                class="MuiTypography-root MuiTypography-caption MuiTypography-noWrap css-1m86euf-MuiTypography-root"
              >
                2024-11-13 08:24:08
              </p>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-7wsi7o"
          data-testid="user-label"
        >
          <div
            class="MuiStack-root css-98m5zp-MuiStack-root"
            data-testid="icon-with-label"
          >
            <div
              class="MuiBox-root css-k008qs"
            >
              <div
                sx="[object Object]"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-r7cbk0-MuiAvatar-root"
                  data-testid="user-avatar"
                >
                  JH
                </div>
              </div>
            </div>
            <div
              class="MuiBox-root css-31l7gp"
            >
              <p
                class="MuiTypography-root MuiTypography-caption MuiTypography-noWrap css-1m86euf-MuiTypography-root"
              >
                James Harper
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
